### PR TITLE
Send files to file system in a driver way

### PR DESCRIPTION
**Pull request by brenoaraujolima**
Depends on #31 

Creating driver structure to send bytes to fyle system.

### CODE REVIEW

File: drivers/drivers.d.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+type DriverBase = {
+    upload_folder: string,
+    limit_file_size: string,
+    makeAuth: (arg0: string) => boolean,
```
Comment (brenoaraujolima) [inline drivers/drivers.d.ts:4]: Return a promise

**Review by uxjp — CHANGES_REQUESTED** [review]
The commit "Removing current writing in api" removes functionality without providing a replacement. It is not atomic. This is a general observation, and since the PR is still in draft, it is expected to have this kind of mistake.

**Review by uxjp — CHANGES_REQUESTED** [review]
It's not explicitly mentioned how to test the code change. An example would be nice.

The title of the PR mentions a feature that we already have. This PR doesn't introduce it but instead makes some other modifications. What exactly are those modifications?

**Comment by brenoaraujolima** [comment]
Ok, @uxjp . I'm working on your notes

**Comment by brenoaraujolima** [comment]
> The commit "Removing current writing in api" removes functionality without providing a replacement. It is not atomic. This is a general observation, and since the PR is still in draft, it is expected to have this kind of mistake.

Fixed

> It says it depends on #31 but its not rebased on #31.

Now it's rebased

> It's not explicitly mentioned how to test the code change. An example would be nice.
The title of the PR mentions a feature that we already have. This PR doesn't introduce it but instead makes some other modifications. What exactly are those modifications?

The change is: sending files to fylesystem in a driver way or, in other words, build the driver for filesystem to make possible send the files using the file system's driver. You're right, title is confused. I'll rewrite the title to make it more clear.

About the test of code change, As shown below, I made available a tenant in database that uses fileSystem driver. 

```json
{
    "tenant": "delriojanderson",
    "driver": "fileSystem",
    "properties": {
        "upload_folder": "uploads",
        "limit_file_size": "4mb"
    }
}

```

So, for test purposes, you can fill your `auth_vars` with [these values](https://gitlab.com/SoftSite/geosales/-/issues/493#note_1404233989) and make the request normally using the scripts in `/client_example` . If everything will be ok, the files will be sent to uploads folder that will be created in runtime.

File: drivers/drivers.d.ts (hunk 1 above)
Comment (brenoaraujolima) [reply drivers/drivers.d.ts]: Fixed

File: api/api.ts [hunk 2]
```diff
@@ -1,9 +1,13 @@
 import express from 'express'
 import * as contentDisposition from 'content-disposition'
-import * as fs from 'fs'
+import bodyParser from 'body-parser'
 import { config as dotenvConfig } from "dotenv-safe"
+import getDocByTenant from '../db/db.js'
+import uploadFile from '../drivers/driver_manager.js'
 
 const app = express()
+app.use(bodyParser.raw({type: 'application/octet-stream', limit : '5mb'}))
```
Comment (uxjp) [inline api/api.ts:9]: 5mb ??

File: api/api.ts [hunk 3]
```diff
@@ -34,22 +38,10 @@ app.post('/upload/:tenant', async (req: express.Request, res: express.Response)
...
-                    console.error(err)
-                    res.status(500).send(`Error saving file ${filename}`)
-                } else {
-                    console.log(`File ${filename} saved`)
-                    res.status(200).send(`File ${filename} saved.`)
-                }
-            })
-        })
+        const tenantConfig = await getDocByTenant(req.params.tenant)
+        const uploadJsonResponse = await uploadFile(tenantConfig, req.body, filename)
```
Comment (uxjp) [inline api/api.ts:43]: `uploadJsonResponse` never used, even if a fail would happen in the upload the `File ${filename} saved` message would be returned.

File: api/api.ts [hunk 4]
```diff
@@ -34,22 +39,10 @@ app.post('/upload/:tenant', async (req: express.Request, res: express.Response)
...
-                if (err) {
-                    console.error(err)
-                    res.status(500).send(`Error saving file ${filename}`)
-                } else {
-                    console.log(`File ${filename} saved`)
-                    res.status(200).send(`File ${filename} saved.`)
-                }
-            })
-        })
+        const tenantConfig = await getDocByTenant(req.params.tenant)
```
Comment (uxjp) [inline api/api.ts:43]: At this point its not necessary, but is important do make a issue to disscuss what may happen if the Mongo DB doens't return any tenant config or any "internet like" problem.

Comment (uxjp) [reply api/api.ts:43]: Also, note that we have coupled a DB structure with the "Uploader Application" and need to provide the code necessary to create the basic structure for the MongoDB database.

File: api/api.ts (hunk 3 above)
Comment (jeffque) [reply api/api.ts:44]: Avoid guiding application to be exception driven. Not a great way to control flow.

It is easy to mess. Prefer doing stuff explicitly. Use a tagged union to deal with multiple possible returns, if it is the case.

Like:

```ts
type Maybe = Err | Success;
type Success = {
  status: “ok”,
  resultField1: typeField1,
  resultField2: typeField2,
    . . .
};
type Err = {
  status: “fail”,
  reason: string,
  errField1: typeField1,
  errField2: typeField2,
    . . .
};
```

Comment (brenoaraujolima) [reply api/api.ts:44]: But this way, I have to check if msg can be send, right?
```ts
if (typeof  uploadJsonResponse === "Success") {
    res.send("file saved successfully")
}
```

File: api/api.ts (hunk 4 above)
Comment (brenoaraujolima) [reply api/api.ts:43]: Issue to discuss these points was open -> #38

**Review by williamvb — APPROVED** [review]

**Pull request merged**